sdk/core/azure_core/src/http/policies/instrumentation/public_api_instrumentation.rs

```diff
@@ -0,0 +1,654 @@
+    }
+
+    /// Creates a span for the public API instrumentation policy.
+    ///
+    /// This function creates a span for the public API instrumentation policy based on the
+    /// public API information in the context.
+    ///
+    /// This function assumes that the `Context` already has a `PublicApiInstrumentationInformation` value,
+    /// if it is not present, it will return `None`.
+    ///
+    /// # Arguments
+    /// - `ctx`: The context containing the public API information.
+    /// - `tracer`: An optional tracer to use for creating the span.
+    ///
+    /// # Returns
+    /// An optional span if the public API information is present and a tracer is available.
+    ///
+    /// If the context already has a span, it will return `None` to avoid nested spans.
+    /// If the context does not have a tracer it will use the value of the `tracer` argument.
+    /// If no tracer can be determined, it will return `None`.
+    ///
+    pub fn create_public_api_span(
+        ctx: &Context,
+        tracer: Option<Arc<dyn Tracer>>,
+    ) -> Option<Arc<dyn Span>> {
+        // If there is a span in the context, we're a nested call, so we just want to forward the request.
+        if ctx.value::<Arc<dyn Span>>().is_some() {
+            trace!("PublicApiPolicy: Nested call detected, forwarding request without instrumentation.");
+            return None;
+        }
+
+        // We next confirm if the context has public API instrumentation information.
+        // Without a public API information, we skip instrumentation.
+        let info = ctx.value::<PublicApiInstrumentationInformation>()?;
+
+        // Get the tracer from either the context or the policy.
+        #[allow(clippy::unnecessary_lazy_evaluations)]
+        let tracer = ctx.value::<Arc<dyn Tracer>>().or_else(|| tracer.as_ref())?;
+
+        // We now have public API information and a tracer.
+        // Calculate the span attributes based on the public API information and
+        // tracer.
+        let mut span_attributes = info
+            .attributes
+            .iter()
+            .map(|attr| {
+                // Convert the attribute to a span attribute.
+                Attribute {
+                    key: attr.key,
+                    value: attr.value.clone(),
+                }
+            })
+            .collect::<Vec<_>>();
+
+        if let Some(namespace) = tracer.namespace() {
+            // If the tracer has a namespace, we set it as an attribute.
+            span_attributes.push(Attribute {
+                key: AZ_NAMESPACE_ATTRIBUTE,
+                value: namespace.into(),
+            });
+        }
+
+        // Create a span with the public API information and attributes.
+        let span =
+            tracer.start_span_with_current(info.api_name, SpanKind::Internal, span_attributes);
+
+        // If nothing is listening to the span, we skip instrumentation.
+        if !span.is_recording() {
+            return None;
```

Comment:
this would be problematic 

1. even if span is not recording (sampled out), we still want to propagate its context. Some of its child spans or logs under it can be recorded and they would all be correlated if current context is set in-process and is propagated downstream
2. there could be some future cases when span start/end callbacks in OTel's span processor got invoked even for non-recording spans, and someone can try to convert spans to metrics this way.

The common way to optimize here is do

- start span with sampling-relevant attributes
- make it current and inject its context regardless of recording
- only add new attributes if span is recording
- end it when the operation is over